src/main/scala/org/broadinstitute/dsde/firecloud/service/StorageService.scala

```diff
@@ -0,0 +1,68 @@
+package org.broadinstitute.dsde.firecloud.service
+
+import akka.actor._
+import akka.pattern._
+import akka.event.Logging
+import org.broadinstitute.dsde.firecloud.{FireCloudExceptionWithErrorReport, Application}
+import org.broadinstitute.dsde.firecloud.dataaccess._
+import org.broadinstitute.dsde.firecloud.model._
+import org.broadinstitute.dsde.firecloud.model.ModelJsonProtocol._
+import org.broadinstitute.dsde.firecloud.service.PerRequest._
+import org.broadinstitute.dsde.firecloud.utils.RestJsonClient
+import spray.http.StatusCodes
+import spray.httpx.SprayJsonSupport._
+import spray.json._
+import spray.json.DefaultJsonProtocol._
+import scala.concurrent.ExecutionContext
+
+/**
+ * Created by mbemis on 11/28/16.
+ */
+object StorageService {
+  sealed trait StorageServiceMessage
+  case class GetObjectStats(bucketName: String, objectName: String) extends StorageServiceMessage
+
+  def props(storageServiceConstructor: UserInfo => StorageService, userInfo: UserInfo): Props = {
+    Props(storageServiceConstructor(userInfo))
+  }
+
+  def constructor(app: Application)(userInfo: UserInfo)(implicit executionContext: ExecutionContext) =
+    new StorageService(userInfo, app.googleServicesDAO)
+}
+
+class StorageService(protected val argUserInfo: UserInfo, val googleServicesDAO: GoogleServicesDAO)(implicit val executionContext: ExecutionContext) extends Actor with RestJsonClient {
+  implicit val system = context.system
+  val log = Logging(system, getClass)
+  implicit val userInfo = argUserInfo
+  import StorageService._
+
+  override def receive: Receive = {
+    case GetObjectStats(bucketName: String, objectName: String) => getObjectStats(bucketName, objectName) pipeTo sender
+  }
+
+  def getObjectStats(bucketName: String, objectName: String) = {
+    requestToObject[JsObject](googleServicesDAO.getObjectMetadataRequest(bucketName, objectName)) flatMap { googleResponse =>
+      googleServicesDAO.fetchPriceList map { googlePrices =>
+        googleResponse.getFields("size").headOption match {
+          case Some(size) => {
+            val fileSizeGB = BigDecimal(size.toString.replaceAll("\"", "")) / Math.pow(1000, 3)
+            val egressPrice = getEgressCost(googlePrices.prices.cpComputeengineInternetEgressNA.tiers, fileSizeGB, 0)
+            RequestComplete(StatusCodes.OK, googleResponse.copy(googleResponse.fields ++ Map("estimatedCostUSD" -> JsNumber(egressPrice))))
+          }
+          case None => RequestComplete(StatusCodes.OK, googleResponse)
+
+        }
+      }
+    }
+  }
+
+  private def getEgressCost(googlePrices: Map[Long, BigDecimal], fileSizeGB: BigDecimal, totalCost: BigDecimal): BigDecimal = {
+    if (fileSizeGB <= 0) totalCost
+    else if(googlePrices.size <= 1) (googlePrices.head._1 * fileSizeGB) + totalCost
```

Comment:
This assumes that the map values are returned in ascending key order, but I'm not sure that there's anything guaranteeing that to be true.

`googlePrices.size` had better not be < 1, otherwise `googlePrices.head._1` (which I think should be `._2`) will give you trouble.

I think you can write this using pattern matching. Something like:
```
googlePrices match {
  case p :: Nil => (p._2 * fileSizeGB) + totalCost
  case p :: t =>
    val (sizeCharged, sizeRemaining) = ...
    getEgressCost(t, sizeRemaining, (sizeCharged * p._2) + totalCost)
}
```

You might even be able to play with the `p` part of the cases to give meaningful names to the parts instead of using `._1` and `._2`. (Disclaimer: I haven't tried to compile this code, so it may need some tweaks.)

I'm not sure what the "92160" key in the JSON price list means, but I guess it's supposed to represent "10+ TB" tier. This implementation using `match` handles that case (assuming that the map values are properly sorted).